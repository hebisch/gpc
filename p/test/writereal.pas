{ Contributed by acahalan@cs.uml.edu under the 2-clause BSD license.

  Slightly modified by Frank Heckenbach: inserted `+1e-24' etc. in
  various places because of rounding inaccuraccies.

  These tests are rather fragile due to the text comparison with
  writereal.out. If this turns out to become a bigger problem, we
  might have to read the values back in Pascal code and compare them
  with some tolerance, see read*.pas. }

PROGRAM giga ( input,output ) ;

var w:Integer;
    p:Integer;
    t:Text;
    s:String (200);
    i:Integer;

begin
   Rewrite (t);

   w := 44;
   p := 0;

   writeln(t, '''', 9.999e9, '''');
   writeln(t, '''', 9.999e9:w, '''');
   writeln(t, '''', 9.999e9:w:p, '''');
   writeln(t, '''', 7.777e9:w:p, '''');
   writeln(t, '''', 9.999e-10, '''');
   writeln(t, '''', 9.999e-10+1e-24:w, '''');
   writeln(t, '''', 9.999e-10:w:p, '''');
   writeln(t, '''', 7.777e-10:w:p, '''');
   writeln(t, '''', 123.456, '''');
   writeln(t, '''', 123.456+1e-12:w, '''');
   writeln(t, '''', 123.456:w:p, '''');
   writeln(t, '''', 0.0001e-9+1e-27, '''');
   writeln(t, '''', 0.0001e-9+1e-27:w, '''');
   writeln(t, '''', 0.0001e-9:w:p, '''');
   writeln(t, '''', 0.0001e-10+1e-28, '''');
   writeln(t, '''', 0.0001e-10+1e-28:w, '''');
   writeln(t, '''', 0.0001e-10:w:p, '''');
   writeln(t, '''', -9.999e9, '''');
   writeln(t, '''', -9.999e9:w, '''');
   writeln(t, '''', -9.999e9:w:p, '''');
   writeln(t, '''', -7.777e9:w:p, '''');
   writeln(t, '''', -9.999e-10, '''');
   writeln(t, '''', -9.999e-10-1e-24:w, '''');
   writeln(t, '''', -9.999e-10:w:p, '''');
   writeln(t, '''', -7.777e-10:w:p, '''');
   writeln(t, '''', -123.456, '''');
   writeln(t, '''', -123.456-1e-12:w, '''');
   writeln(t, '''', -123.456:w:p, '''');
   writeln(t, '''', -0.0001e-9-1e-27, '''');
   writeln(t, '''', -0.0001e-9-1e-27:w, '''');
   writeln(t, '''', -0.0001e-9:w:p, '''');
   writeln(t, '''', -0.0001e-10-1e-28, '''');
   writeln(t, '''', -0.0001e-10-1e-28:w, '''');
   writeln(t, '''', -0.0001e-10:w:p, '''');

{$if False}  { Can crash on non-IEEE FP implementations }
   writeln(t, '''', (-0.0)/0.0:w:p, '''');
   writeln(t, '''',   0.0 /0.0:w:p, '''');
   writeln(t, '''', (-1.0)/0.0:w:p, '''');
   writeln(t, '''',   1.0 /0.0:w:p, '''');
{$endif}
   writeln(t, '''', (-0.0)/1.0:w:p, '''');
   writeln(t, '''',   0.0 /1.0:w:p, '''');
   writeln(t, '''', (-1.0)/1.0:w:p, '''');
   writeln(t, '''',   1.0 /1.0:w:p, '''');
{$if False}  { Can crash on non-IEEE FP implementations }
   writeln(t, '''', (-0.0)/(-0.0):w:p, '''');
   writeln(t, '''',   0.0 /(-0.0):w:p, '''');
   writeln(t, '''', (-1.0)/(-0.0):w:p, '''');
   writeln(t, '''',   1.0 /(-0.0):w:p, '''');
{$endif}
   writeln(t, '''', (-0.0)/(-1.0):w:p, '''');
   writeln(t, '''',   0.0 /(-1.0):w:p, '''');
   writeln(t, '''', (-1.0)/(-1.0):w:p, '''');
   writeln(t, '''',   1.0 /(-1.0):w:p, '''');

   writeln(t);

   { Compensate for differences in accurary and rounding between systems. }
   Reset (t);
   while not EOF (t) do
     begin
       ReadLn (t, s);
       i := 18;
       while (i <= Length (s)) and (s[i] in ['0' .. '9']) do
         begin
           s[i] := '0';
           Inc (i)
         end;
       writeln (s)
     end

end.
