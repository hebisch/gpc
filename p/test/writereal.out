' 9.999000000000000e+09'
' 9.9990000000000000000000000000000000000e+09'
'                                  9999000000'
'                                  7777000000'
' 9.999000000000000e-10'
' 9.9990000000000000000000000000000000000e-10'
'                                           0'
'                                           0'
' 1.234560000000000e+02'
' 1.2345600000000000000000000000000000000e+02'
'                                         123'
' 1.000000000000000e-13'
' 1.0000000000000000000000000000000000000e-13'
'                                           0'
' 1.000000000000000e-14'
' 1.0000000000000000000000000000000000000e-14'
'                                           0'
'-9.999000000000000e+09'
'-9.9990000000000000000000000000000000000e+09'
'                                 -9999000000'
'                                 -7777000000'
'-9.999000000000000e-10'
'-9.9990000000000000000000000000000000000e-10'
'                                          -0'
'                                          -0'
'-1.234560000000000e+02'
'-1.2345600000000000000000000000000000000e+02'
'                                        -123'
'-1.000000000000000e-13'
'-1.0000000000000000000000000000000000000e-13'
'                                          -0'
'-1.000000000000000e-14'
'-1.0000000000000000000000000000000000000e-14'
'                                          -0'
'                                           0'
'                                           0'
'                                          -1'
'                                           1'
'                                           0'
'                                           0'
'                                           1'
'                                          -1'

